File: lib/services/api_client.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:http/http.dart' as http;

import '../data/models.dart';

/// Matches the `demo-user` seeded by the backend's `npm run seed` script
/// (`src/data/seed.ts`). There's no auth in this wireframe, so every device
/// shares this identity.
const String demoUserId = 'demo-user';

class ApiException implements Exception {
  final String message;
  ApiException(this.message);

  @override
  String toString() => 'ApiException: $message';
}

/// The result of starting a Learn session: the backend session id plus the
/// `ChallengeDecision` JSON (`src/core/types/visual.ts`) it recommended for
/// the scanned scene, if any.
class LearnSessionResult {
  final String sessionId;
  final Map<String, dynamic>? decision;
  final LearningContent learningContent;
  const LearnSessionResult({
    required this.sessionId,
    required this.learningContent,
    this.decision,
  });
}

/// Thin client for the Node/Express backend in this repo (`src/server.ts`).
/// Every call can throw [ApiException] or a network error — callers should
/// catch and fall back to local/mock behavior so the wireframe still works
/// with the backend offline.
class ApiClient {
  ApiClient._();
  static final ApiClient _instance = ApiClient._();
  factory ApiClient() => _instance;

  /// An explicit override, e.g. `--dart-define=API_BASE_URL=https://foo.trycloudflare.com`
  /// — needed when the app and backend are served from different origins,
  /// such as two separate tunnels, where the app can't infer the backend's
  /// address from its own URL.
  static const String _override = String.fromEnvironment('API_BASE_URL');

  /// Backend always runs on port 4000 (see `.env`) unless [_override] is set.
  /// On web, it's normally served from the same host as this build
  /// (`localhost` or a LAN IP), read off the page's own URL. On a native
  /// build (iOS Simulator, physical device), there's no page URL to read —
  /// the iOS Simulator shares the Mac's network stack, so `localhost`
  /// reaches a backend running on the host machine directly.
  static String get baseUrl {
    if (_override.isNotEmpty) return _override;
    final host = kIsWeb
        ? (Uri.base.host.isNotEmpty ? Uri.base.host : 'localhost')
        : 'localhost';
    return 'http://$host:4000';
  }

  Uri _uri(String path) => Uri.parse('$baseUrl$path');

  Map<String, dynamic> _unwrap(http.Response res) {
    final decoded = jsonDecode(res.body) as Map<String, dynamic>;
    if (res.statusCode >= 400 || decoded['success'] == false) {
      final message =
          (decoded['error'] as Map?)?['message']?.toString() ??
          'Request failed (${res.statusCode})';
      throw ApiException(message);
    }
    return decoded;
  }

  Future<Map<String, dynamic>> _post(
    String path,
    Map<String, dynamic> body, {
    Duration timeout = const Duration(seconds: 8),
  }) async {
    final res = await http
        .post(
          _uri(path),
          headers: const {'Content-Type': 'application/json'},
          body: jsonEncode(body),
        )
        .timeout(timeout);
    final decoded = _unwrap(res);
    return decoded['data'] as Map<String, dynamic>? ?? {};
  }

  Future<List<dynamic>> _getList(String path) async {
    final res = await http.get(_uri(path)).timeout(const Duration(seconds: 8));
    final decoded = _unwrap(res);
    return decoded['data'] as List<dynamic>? ?? [];
  }

  /// Uploads a captured photo to the backend's OpenCV pipeline
  /// (`computer_vision/scene_analysis.py`) and returns the raw SceneAnalysis
  /// JSON (colors/shapes/textures/lines/patterns) it detected.
  Future<Map<String, dynamic>> analyzeScan(Uint8List photoBytes) async {
    final res = await http
        .post(
          _uri('/api/scan/analyze'),
          headers: const {'Content-Type': 'image/jpeg'},
          body: photoBytes,
        )
        .timeout(const Duration(seconds: 20));
    final decoded = _unwrap(res);
    return decoded['data'] as Map<String, dynamic>? ?? {};
  }

  Future<String> createSession({
    required String userId,
    required String mode,
  }) async {
    final data = await _post('/api/sessions', {'userId': userId, 'mode': mode});
    return data['id'] as String;
  }

  /// Posts the (currently mocked) scene analysis and returns the backend's
  /// `ChallengeDecision`, or null if the scene had insufficient features.
  Future<Map<String, dynamic>?> postSceneAnalysis({
    required String sessionId,
    required Map<String, dynamic> scene,
  }) async {
    final data = await _post('/api/sessions/$sessionId/scene-analysis', scene);
    final recommendation = data['recommendation'] as Map<String, dynamic>?;
    return recommendation?['decision'] as Map<String, dynamic>?;
  }

  Future<Map<String, dynamic>> createChallengeInstance({
    required String sessionId,
    required String title,
    required String instructions,
  }) {
    return _post('/api/sessions/$sessionId/challenge-instance', {
      'title': title,
      'instructions': instructions,
    });
  }

  Future<Map<String, dynamic>> generateCreativeChallenge(
    String sessionId, {
    Map<String, dynamic>? learningContext,
  }) {
    return _post('/api/sessions/$sessionId/creative-challenge', {
      if (learningContext != null) 'learningContext': learningContext,
    }, timeout: const Duration(seconds: 30));
  }

  Future<LearningContent> generateLearningContent(String sessionId) async {
    final data = await _post('/api/learning/content', {
      'sessionId': sessionId,
    }, timeout: const Duration(seconds: 30));
    return LearningContent.fromJson(data);
  }

  Future<List<Map<String, dynamic>>> getChallengeHistory(String userId) async {
    final list = await _getList('/api/users/$userId/challenge-instances');
    return list.cast<Map<String, dynamic>>();
  }

  Future<Map<String, dynamic>?> recommendChallenge({
    required Map<String, dynamic> scene,
    required String userId,
  }) async {
    final data = await _post('/api/challenges/recommend', {
      'sceneAnalysis': scene,
      'userId': userId,
    });
    return data['decision'] as Map<String, dynamic>?;
  }

  Future<void> completeSession({
    required String sessionId,
    required String userId,
    String? challengeType,
    Map<String, dynamic>? metadata,
  }) async {
    await _post('/api/sessions/$sessionId/complete', {
      'userId': userId,
      if (challengeType != null) 'challengeType': challengeType,
      if (metadata != null) 'artworkMetadata': metadata,
    });
  }

  Future<List<Map<String, dynamic>>> getArtworks(String userId) async {
    final list = await _getList('/api/users/$userId/artworks');
    return list.cast<Map<String, dynamic>>();
  }
}

File: web-next/src/lib/apiClient.ts
// Thin client for the existing Node/Express backend (src/server.ts).
// Same endpoints, same request/response shapes as lib/services/api_client.dart —
// only the transport (fetch vs. package:http) and language changed.

import type { CreativeChallenge, LearningContent } from './types';
import { creativeChallengeFromInstance } from './types';

export const demoUserId = 'demo-user';

export class ApiException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ApiException';
  }
}

/**
 * Matches ApiClient.baseUrl in lib/services/api_client.dart: an explicit
 * NEXT_PUBLIC_API_URL override (needed when frontend/backend are on
 * different origins, e.g. a tunnel), else the page's own host on port 4000.
 */
export function apiBaseUrl(): string {
  const override = process.env.NEXT_PUBLIC_API_URL;
  if (override) return override;
  if (typeof window !== 'undefined' && window.location.hostname) {
    return `http://${window.location.hostname}:4000`;
  }
  return 'http://localhost:4000';
}

async function unwrap(res: Response): Promise<Record<string, unknown>> {
  const decoded = (await res.json()) as Record<string, unknown>;
  if (!res.ok || decoded.success === false) {
    const err = decoded.error as { message?: string } | undefined;
    throw new ApiException(err?.message ?? `Request failed (${res.status})`);
  }
  return decoded;
}

async function post(path: string, body: unknown, timeoutMs = 8000): Promise<Record<string, unknown>> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(`${apiBaseUrl()}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    const decoded = await unwrap(res);
    return (decoded.data as Record<string, unknown>) ?? {};
  } finally {
    clearTimeout(timer);
  }
}

async function getList(path: string): Promise<Record<string, unknown>[]> {
  const res = await fetch(`${apiBaseUrl()}${path}`);
  const decoded = await unwrap(res);
  return (decoded.data as Record<string, unknown>[]) ?? [];
}

export async function createSession(userId: string, mode: string): Promise<string> {
  const data = await post('/api/sessions', { userId, mode });
  return data.id as string;
}

export async function postSceneAnalysis(
  sessionId: string,
  scene: unknown
): Promise<Record<string, unknown> | null> {
  const data = await post(`/api/sessions/${sessionId}/scene-analysis`, scene);
  const recommendation = data.recommendation as Record<string, unknown> | undefined;
  return (recommendation?.decision as Record<string, unknown> | undefined) ?? null;
}

export async function generateCreativeChallenge(
  sessionId: string,
  learningContext?: Record<string, unknown>
): Promise<CreativeChallenge> {
  const data = await post(
    `/api/sessions/${sessionId}/creative-challenge`,
    learningContext ? { learningContext } : {},
    30000
  );
  return creativeChallengeFromInstance(data);
}

export async function generateLearningContent(sessionId: string): Promise<LearningContent> {
  const data = await post('/api/learning/content', { sessionId }, 30000);
  return data as unknown as LearningContent;
}

export async function completeSession(
  sessionId: string,
  userId: string,
  challengeType?: string,
  metadata?: Record<string, unknown>
): Promise<void> {
  await post(`/api/sessions/${sessionId}/complete`, {
    userId,
    ...(challengeType ? { challengeType } : {}),
    ...(metadata ? { artworkMetadata: metadata } : {}),
  });
}

export async function getArtworks(userId: string): Promise<Record<string, unknown>[]> {
  return getList(`/api/users/${userId}/artworks`);
}
